lib/Component.spec.lua

```diff
@@ -173,4 +173,56 @@ return function()
 
 		Reconciler.teardown(instance)
 	end)
+
+	describe("setState", function()
+		it("should throw when called in init or render", function()
+			local InitComponent = Component:extend("InitComponent")
+
+			function InitComponent:init()
+				self:setState({
+					a = 1
+				})
+			end
+
+			local RenderComponent = Component:extend("RenderComponent")
+
+			function RenderComponent:render()
+				self:setState({
+					a = 1
+				})
+			end
+
+			local initElement = Core.createElement(InitComponent)
+			local renderElement = Core.createElement(RenderComponent)
+
+			expect(function()
+				Reconciler.reify(initElement)
+			end).to.throw()
+
+			expect(function()
+				Reconciler.reify(renderElement)
+			end).to.throw()
+		end)
```

Comment:
I think we can break this one into two tests, just like how `willUnmount` is a separate test